### ISSUE TITLE
Implement hash-index-based join optimization

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4802]
This task is automatically imported from the old Task Issue Board and it was originally created by Radosław Waśko.
Original issue is [here](https://www.pivotaltracker.com/story/show/183913276).
<hr />

Currently, Join is based on a naive full scan. This is slow and can be improved.

Implement a hashmap based optimization for both Equals and Equals_Ignore_Case (with special case-folded Text hashing). It should then delegate to a Scan based solution for remaining Between cases.

The optimization will currently only work for builtin types for which we have a well-defined hashing notion. For other types it will also rely on Scan based.

The optimization should take all fields on which it can perform efficiently to perform an initial hash-based pass and then fallback to a scan on filtered groups (which will hopefully be significantly smaller).

It does not need particular preparation for merging it with the Range-based join for efficient Between - this will be handled as part of another task.

## Comments:
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-12-09):

**Progress:** Found some issues with table construction allowing invalid arguments - fixed that and added tests for these cases on my current WIP branch. Spotted a super weird and hard to reproduce bug when matching `Object[]` with `_ : Array`. Was trying to figure out a sensible repro to send it towards the engine team - but after all finding a repro got me so deep into debugging that I just found the root cause and fixed the bug (fix pending on my branch). Created a suite of performance tests that will be ensuring that our joins do not rely on O(N*M) algorithms - the tests will only pass if our performance is close to linear O(|output size|). Created a test for Between for the future task using the same pattern although it may need some tweaking as Between will be N log N. It should be finished by 2022-12-15.

**Next Day:** Next day I will be working on the [same task](https://www.pivotaltracker.com/story/show/183913276). Integrate a join strategy that will still use an index for mixed join conditions. See if we can reuse MultiValueIndex. If so, tweak MultiValueIndex to support per-column selection of case insensitive text folding. (Enso Bot - Dec 10, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for today (2022-12-12):

**Progress:** Integrating the index-strategy with the scan-based fallback. Got a basic implementation; TODO integrate it with case-insensitive equality. Prepared an external PR for integration, added a prototype of a meta-test suite for testing our test suite. It should be finished by 2022-12-15.

**Next Day:** Next day I will be working on the [same task](https://www.pivotaltracker.com/story/show/183913276). Tweak MultiValueIndex to support per-column selection of case insensitive text folding to get full support for hash-based joins where possible. Add one more perf. test case and see if the test suite is complete enough, prepare the PR for review. (Enso Bot - Dec 12, 2022)
<hr />
**Radosław Waśko** reports a new **STANDUP** for yesterday (2022-12-13):

**Progress:** Implemented case insensitive indexing, refactored MultiValueIndex to make it more type safe and make it more explicit whether the index uses ordered (compare_to) or unordered (hashing) maps. Submitted the PR for review. It should be finished by 2022-12-15.

**Next Day:** Next day I will be working on the [#184032869 task](https://www.pivotaltracker.com/story/show/184032869). Get the PR merged. Start work on Database.join. Add some relevant test cases, start generating simple SQL joins - then later will be working towards the reaching feature parity with in-memory implementation. (Enso Bot - Dec 14, 2022)
<hr />
